File: functions/predict.py
from http.server import HTTPServer, BaseHTTPRequestHandler
import json
import os
import numpy as np
from joblib import load
import traceback

# Load model and scaler
current_dir = os.path.dirname(os.path.abspath(__file__))
model = load(os.path.join(current_dir, '../model/model.pkl'))
scaler = load(os.path.join(current_dir, '../model/scaler.pkl'))

def calculate_soil_health_score(nitrogen, phosphorous, potassium, ph):
    """Calculate soil health score based on nutrient levels and pH"""
    # Define ideal ranges
    ideal_ranges = {
        'N': (20, 140),
        'P': (10, 50),
        'K': (20, 200),
        'pH': (6.0, 7.5)
    }
    
    # Calculate individual scores
    n_score = 100 - min(100, abs(nitrogen - np.mean(ideal_ranges['N'])) / (ideal_ranges['N'][1] - ideal_ranges['N'][0]) * 100)
    p_score = 100 - min(100, abs(phosphorous - np.mean(ideal_ranges['P'])) / (ideal_ranges['P'][1] - ideal_ranges['P'][0]) * 100)
    k_score = 100 - min(100, abs(potassium - np.mean(ideal_ranges['K'])) / (ideal_ranges['K'][1] - ideal_ranges['K'][0]) * 100)
    ph_score = 100 - min(100, abs(ph - np.mean(ideal_ranges['pH'])) / (ideal_ranges['pH'][1] - ideal_ranges['pH'][0]) * 100)
    
    # Calculate weighted average
    return float(n_score * 0.25 + p_score * 0.25 + k_score * 0.25 + ph_score * 0.25)

def get_recommendations(soil_health_score, ph):
    """Generate additional recommendations based on soil health score and pH"""
    recommendations = []
    
    if soil_health_score < 60:
        recommendations.append("Soil health needs improvement.")
        if ph < 6.0:
            recommendations.append("Consider adding lime to increase soil pH.")
        elif ph > 7.5:
            recommendations.append("Consider adding sulfur to decrease soil pH.")
    
    if soil_health_score < 40:
        recommendations.append("Urgent soil treatment recommended.")
    elif soil_health_score < 70:
        recommendations.append("Regular soil monitoring advised.")
    
    return " ".join(recommendations) if recommendations else "Soil conditions are generally good."

def handler(event, context):
    """Netlify Function handler"""
    # Only allow POST requests
    if event['httpMethod'] != 'POST':
        return {
            'statusCode': 405,
            'body': json.dumps({'error': 'Method not allowed'})
        }

    try:
        # Parse request body
        body = json.loads(event['body'])
        
        # Extract parameters
        nitrogen = float(body.get('nitrogen', 0))
        phosphorous = float(body.get('phosphorous', 0))
        potassium = float(body.get('potassium', 0))
        temperature = float(body.get('temperature', 0))
        humidity = float(body.get('humidity', 0))
        ph = float(body.get('ph', 0))
        rainfall = float(body.get('rainfall', 200))  # Default value

        # Prepare input data
        input_data = np.array([[
            nitrogen, phosphorous, potassium,
            temperature, humidity, ph, rainfall
        ]])
        
        # Scale the input
        input_scaled = scaler.transform(input_data)
        
        # Get prediction probabilities
        probabilities = model.predict_proba(input_scaled)[0]
        
        # Get top 3 predictions
        top_indices = np.argsort(probabilities)[-3:][::-1]
        recommended_crops = [model.classes_[i] for i in top_indices]
        confidence_scores = [float(probabilities[i]) for i in top_indices]
        
        # Calculate soil health score
        soil_health_score = calculate_soil_health_score(
            nitrogen, phosphorous, potassium, ph
        )
        
        # Get additional recommendations
        additional_recommendations = get_recommendations(soil_health_score, ph)
        
        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Headers': 'Content-Type',
                'Access-Control-Allow-Methods': 'POST, OPTIONS'
            },
            'body': json.dumps({
                'recommended_crops': recommended_crops,
                'confidence_scores': confidence_scores,
                'soil_health_score': soil_health_score,
                'additional_recommendations': additional_recommendations
            })
        }
        
    except Exception as e:
        print(f"Prediction error: {str(e)}")
        print(traceback.format_exc())
        return {
            'statusCode': 500,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps({
                'error': 'Failed to make prediction',
                'details': str(e)
            })
        } 

File: functions/predict.js
const { joblib } = require('@scikit-learn/joblib');
const path = require('path');

// Load model and scaler
const modelPath = path.join(__dirname, '../model/model.pkl');
const scalerPath = path.join(__dirname, '../model/scaler.pkl');

let model = null;
let scaler = null;

try {
  model = joblib.load(modelPath);
  scaler = joblib.load(scalerPath);
} catch (error) {
  console.error('Error loading model:', error);
}

function calculateSoilHealthScore(nitrogen, phosphorous, potassium, ph) {
  // Define ideal ranges
  const idealRanges = {
    N: [20, 140],
    P: [10, 50],
    K: [20, 200],
    pH: [6.0, 7.5]
  };
  
  // Calculate individual scores
  const nScore = 100 - Math.min(100, Math.abs(nitrogen - ((idealRanges.N[1] + idealRanges.N[0]) / 2)) / (idealRanges.N[1] - idealRanges.N[0]) * 100);
  const pScore = 100 - Math.min(100, Math.abs(phosphorous - ((idealRanges.P[1] + idealRanges.P[0]) / 2)) / (idealRanges.P[1] - idealRanges.P[0]) * 100);
  const kScore = 100 - Math.min(100, Math.abs(potassium - ((idealRanges.K[1] + idealRanges.K[0]) / 2)) / (idealRanges.K[1] - idealRanges.K[0]) * 100);
  const phScore = 100 - Math.min(100, Math.abs(ph - ((idealRanges.pH[1] + idealRanges.pH[0]) / 2)) / (idealRanges.pH[1] - idealRanges.pH[0]) * 100);
  
  // Calculate weighted average
  return (nScore * 0.25 + pScore * 0.25 + kScore * 0.25 + phScore * 0.25);
}

function getRecommendations(soilHealthScore, ph) {
  const recommendations = [];
  
  if (soilHealthScore < 60) {
    recommendations.push("Soil health needs improvement.");
    if (ph < 6.0) {
      recommendations.push("Consider adding lime to increase soil pH.");
    } else if (ph > 7.5) {
      recommendations.push("Consider adding sulfur to decrease soil pH.");
    }
  }
  
  if (soilHealthScore < 40) {
    recommendations.push("Urgent soil treatment recommended.");
  } else if (soilHealthScore < 70) {
    recommendations.push("Regular soil monitoring advised.");
  }
  
  return recommendations.length ? recommendations.join(" ") : "Soil conditions are generally good.";
}

exports.handler = async (event, context) => {
  // Only allow POST requests
  if (event.httpMethod !== 'POST') {
    return {
      statusCode: 405,
      body: JSON.stringify({ error: 'Method not allowed' })
    };
  }

  try {
    const data = JSON.parse(event.body);
    const {
      nitrogen,
      phosphorous,
      potassium,
      temperature,
      humidity,
      ph,
      rainfall = 200 // Default value
    } = data;

    // Validate input data
    if (!model || !scaler) {
      throw new Error('Model not loaded');
    }

    // Prepare input data
    const inputData = [[
      nitrogen,
      phosphorous,
      potassium,
      temperature,
      humidity,
      ph,
      rainfall
    ]];

    // Scale the input
    const inputScaled = scaler.transform(inputData);

    // Get prediction probabilities
    const probabilities = model.predictProba(inputScaled)[0];

    // Get top 3 predictions
    const topIndices = Array.from(probabilities)
      .map((prob, index) => ({ prob, index }))
      .sort((a, b) => b.prob - a.prob)
      .slice(0, 3);

    const recommendedCrops = topIndices.map(({ index }) => model.classes_[index]);
    const confidenceScores = topIndices.map(({ prob }) => prob);

    // Calculate soil health score
    const soilHealthScore = calculateSoilHealthScore(nitrogen, phosphorous, potassium, ph);

    // Get additional recommendations
    const additionalRecommendations = getRecommendations(soilHealthScore, ph);

    return {
      statusCode: 200,
      headers: {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Content-Type',
        'Access-Control-Allow-Methods': 'POST, OPTIONS'
      },
      body: JSON.stringify({
        recommended_crops: recommendedCrops,
        confidence_scores: confidenceScores,
        soil_health_score: soilHealthScore,
        additional_recommendations: additionalRecommendations
      })
    };
  } catch (error) {
    console.error('Prediction error:', error);
    return {
      statusCode: 500,
      headers: {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*'
      },
      body: JSON.stringify({
        error: 'Failed to make prediction',
        details: error.message
      })
    };
  }
}; 